src/lib/crypto/kimchi_backend/gadgets/ec_group.ml

```diff
@@ -0,0 +1,3965 @@
+        @@ Foreign_field.Element.Standard.to_limbs @@ Affine.x sum ;
+        Foreign_field.External_checks.append_bound_check external_checks
+        @@ Foreign_field.Element.Standard.to_limbs @@ Affine.y sum ;
+
+        (* Group double: double_base = base + base *)
+        let double_base =
+          match doubles with
+          | None ->
+              let double_base =
+                double (module Circuit) external_checks curve base
+              in
+              (* Bounds 2:
+               *   Input is previous result, so already checked.
+               *   Initial base is gadget inputs and checked externally.
+               *   Result bounds check below.
+               *)
+              Foreign_field.External_checks.append_bound_check external_checks
+              @@ Foreign_field.Element.Standard.to_limbs @@ Affine.x double_base ;
+              Foreign_field.External_checks.append_bound_check external_checks
+              @@ Foreign_field.Element.Standard.to_limbs @@ Affine.y double_base ;
+              double_base
+          | Some doubles ->
+              (* When the base point is public (e.g. the secp256k1 generator) we can
+               * improve performance by having them as precomputed public parameters *)
+              doubles.(i)
+        in
+
+        (* Group add conditionally *)
+        let acc = Affine.if_ (module Circuit) bit ~then_:sum ~else_:acc in
+
+        (acc, double_base) )
+  in
+
+  (* Subtract init_point from accumulator for final result *)
+  add (module Circuit) external_checks curve acc curve.ia.neg_acc
+
+(* Gadget to check point is in the subgroup
+ *   nP = O
+ * where n is the elliptic curve group order and O is the point at infinity
+ *)
+let check_subgroup (type f)
+    (module Circuit : Snark_intf.Run with type field = f)
+    (external_checks : f Foreign_field.External_checks.t)
+    (curve : f Curve_params.InCircuit.t) ?(doubles : f Affine.t array option)
+    (point : f Affine.t) =
+  (* Subgroup check: nP = O
+   *   We don't support identity element, so instead we check
+   *     ((n - 1) + 1)P = O
+   *     (n - 1)P = -P
+   *)
+
+  (* C1: Compute (n - 1)P *)
+  let n_minus_one_point =
+    scalar_mul
+      (module Circuit)
+      external_checks curve ?doubles curve.order_minus_one_bits point
+  in
+  (* Bounds 1: Left input is public constant (no bounds check required)
+   *           Right input is gadget input (checked externally)
+   *           Result bound check below
+   *)
+  Foreign_field.External_checks.append_bound_check external_checks
+  @@ Foreign_field.Element.Standard.to_limbs @@ Affine.x n_minus_one_point ;
+  Foreign_field.External_checks.append_bound_check external_checks
+  @@ Foreign_field.Element.Standard.to_limbs @@ Affine.y n_minus_one_point ;
+
+  (* C2: Compute -P *)
+  let minus_point = negate (module Circuit) curve point in
+  (* Result row *)
+  Foreign_field.result_row (module Circuit) ~label:"minus_point_y"
```

Comment:
I don't love having to remember to include the result row manually 😐

Reply:
You don't always have to waste a row when negating if the value negated is already part of the witness.  In this case it is not, so we need to add it.